design/ceph/resource-dependencies.md

```diff
@@ -0,0 +1,259 @@
+compelling use-case for treating StorageClasses as dependencies, this functionality can be
+implemented at that time.
+
+### Detecting object bucket dependents without querying lib-bucket-provisioner or COSI resources
+Detecting lib-bucket-provisioner ObjectBuckets that are dependent on a given CephObjectStore
+requires inspecting ObjectBuckets for a reference to a bucket found in the object store as well as a
+reference to the address of the Kubernetes Service created for access to RGWs
+(`<service name>.<namespace>.svc`). Detecting COSI Buckets will be similar.
+
+This detection does require accessing external APIs (lib-bucket-provisioner and COSI). This is
+non-ideal for COSI whose APIs will be progressing from v1alpha1, through beta stages, and then into
+v1 in the coming months/years. Rook can merely check for the existence of buckets in order to
+support lib-bucket-provisioner and COSI simultaneously with the same code. This would also remove
+any need for Rook to update its COSI API for dependency checking though it will still need to
+update its API to continue operating as a COSI driver. This would allow Rook to block deletion of
+CephObjectStores without care for who has "claimed" a bucket.
+
+Given that a Kubernetes cluster might have many hundreds of ObjectBuckets, or COSI Buckets, having a
+simpler way of querying for dependents can lighten the load on the Kubernetes API server and reduce
+Rook's resource usage.
+
+Since OBs and COSI Buckets both result in buckets being created in a CephObjectStore, Rook can query
+the object store and block deletion if buckets exist.
+
+This is an elegant solution for blocking when buckets are claimed by these outside resources. It
+does mean that CephObjectStores that have had buckets created in them directly (by admins or users)
+will block until the buckets are manually deleted. An admin may need to request that users remove
+unneeded buckets or may instead remove the buckets themselves.
+
+If the admin wishes to preserve the CephObjectStore's pools on deletion along with their data, the
+admin may merely remove the finalizer on the CephObjectStore.
+
+While this requires more steps to delete the CephObjectStore, it provides additional safety for user
+data by requiring users or admins to inspect and clean up unneeded buckets.
+
+The main downside to this approach is that Rook will not be able to report if there are specific
+OBs or COSI Buckets consuming storage from the CephObjectStore. An admin would need to examine the
+resources in their cluster to determine if there are claims to the storage manually. A diligent
+admin will likely have done this work beforehand.
+
+This document proposes to implement this strategy to avoid reliance on external APIs. Rook
+developers should revisit the decision at a later date to discuss whether the strategy is continuing
+to adequately meet users' needs and whether the drawbacks noted are causing any issues.
+
+### Implementation in stages
```

Comment:
I think the summary, in my mind, is the proposed dependency relationships and the diagram. This is merely a detail about when things will be implemented.

Reply:
There are a lot of detailed dependencies and we aren't planning to enforce them all, right? For me at least, it's helpful to have a "tldr" section at the top on what the proposal is before diving into the details. But as long as we have this section somewhere in the doc, I'll be content. :)

Reply:
In my estimation, we should either enforce them all, or only enforce dependencies from the `CephCluster`. Otherwise, users may be confused when they are protected from some dependencies but not others, and it will be difficult to document the minutiae of what is and what is not protected. 

If it is all protected, then we can just document that Rook resources should be protected from being deleted if there are other Rook resources referencing them, and if they aren't it can be considered a bug. Additional protections include buckets created by users, lib-bucket-provisioner OBCs, or COSI Buckets as well as PersistentVolumes that reference Rook storage.

Reply:
I created a new summary section at the top.